hotp/ta/hotp_ta.c

```diff
@@ -0,0 +1,277 @@
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+ * AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+ * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+ * ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE
+ * LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+ * CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+ * SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+ * INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+ * CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+ * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ * POSSIBILITY OF SUCH DAMAGE.
+ */
+#include <hotp_ta.h>
+#include <string.h>
+#include <tee_internal_api_extensions.h>
+#include <tee_internal_api.h>
+
+/*
+ * This macro checks whether the return value is not equal to TEE_SUCCESS, if it
+ * is not equal, then it will set the return value to 'res' and then goto
+ * 'label'.
+ */
+#define CHECK_EXIT(label, res, ret_val) \
+	if (ret_val != TEE_SUCCESS) { \
+		EMSG("0x%08x", ret_val); \
+		res = ret_val; \
+		goto label; }
+
+/*
+ * This macro checks the expected parameters, if they are not the same, then it
+ * will set 'res' to TEE_ERROR_BAD_PARAMETERS and will goto label 'label'.
+ */
+#define CHECK_EXP_PARAM_EXIT(label, res, pt, ept) \
+	if (pt != ept) { \
+		EMSG("Expected: 0x%x, got: 0x%x", ept, pt); \
+		res = TEE_ERROR_BAD_PARAMETERS; \
+		goto label; }
+
+#define SWAP64(v) \
+	(((uint64_t)(v) << 56) | \
+	 (((uint64_t)(v) & 0x000000000000FF00) << 40) | \
+	 (((uint64_t)(v) & 0x0000000000FF0000) << 24) | \
+	 (((uint64_t)(v) & 0x00000000FF000000) << 8) | \
+	 (((uint64_t)(v) & 0x000000FF00000000) >> 8) | \
+	 (((uint64_t)(v) & 0x0000FF0000000000) >> 24) | \
+	 (((uint64_t)(v) & 0x00FF000000000000) >> 40) | \
+	 ((uint64_t)(v) >> 56))
+
+/* The size of a SHA1 hash in bytes. */
+#define SHA1_HASH_SIZE 20
+
+/* GP says that for HMAC SHA-1, max is 512 bits and min 80 bits. */
+#define MAX_KEY_SIZE 64 /* In bytes */
+#define MIN_KEY_SIZE 10 /* In bytes */
+
+/* Dynamic Binary Code 2 Modulu, which is 10^6 according to the spec. */
+#define DBC2_MODULO 1000000
+
+/*
+ * Currently this only supports a single key, in the future this could be
+ * updated to support multiple users, all with different unique keys (stored
+ * using secure storage).
+ */
+static uint8_t K[MAX_KEY_SIZE];
+static uint32_t K_len;
+
+/*
+ * The HOTP counter should be a 8 byte array where byte 0 is the MSB and byte 7
+ * is the LSB. To make it easier to update the variable, just simply use a
+ * uint64_t instead. Since the HOTP counter needs to be in big endian, we need
+ * to swap it when incrementing.
+ */
+static uint64_t counter_be = 0;
+
+static void inc_counter(uint64_t *c_be)
+{
+	uint64_t c_le = SWAP64(*c_be);
```

Comment:
Byteswapping is fun, but I'd rather see a conversion from 64-bit Big Endian to native endian.

Reply:
;-)

Reply:
Are you looking for something like `__be64_to_cpup` in Linux kernel? If not I don't follow. Anyhow, I've gone back to how I did it in earlier commits, a bit more magic, but less code any no issues with endianness.